docs/tutorial.md

```diff
@@ -323,3 +323,449 @@ as well as the C++ standard library and increases the dependency surface.
+`g_usleep` or a delay in seconds that shows a progress bar with
+`fu_device_sleep_with_progress`. Note that, in both cases, this will
+stop the application main loop during the wait, so use it only when
+necessary.
+
+### How to define private flags
+
+Besides the regular flags and internal flags that any device can have, a
+device can define private flags for specific uses. These can be enabled
+in the code as well as in quirk files, just as the rest of flags. To
+define a private flag:
+
+1. Define the flag value. This is normally defined as a macro that
+expands to a binary flag, for example: `#define MY_PRIVATE_FLAG (1 << 2)`.
+Note that this will be part of the ABI, so it must be versioned.
+
+2. Call `fu_device_register_private_flag` in the device init function
+and assign a string identifier to the flag: `fu_device_register_private_flag (FU_DEVICE (self), MY_PRIVATE_FLAG, "myflag");`
+
+You can then add it to the device programmatically with
+`fu_device_add_private_flag`, remove it with `fu_device_remove_private_flag`
+and query it with `fu_device_has_private_flag`. In a quirk file, you can
+add the flag identifier to the Flags attribute of a device (eg. `Flags =
+myflag,is-bootloader`)
+
+### How to make fwupd wait for a device replug
+
+Certain devices require a disconnection and reconnection to start the
+update process. A common example are devices that have two booting
+modes: application or runtime mode, and bootloader mode, where the
+runtime mode is the normal operation mode and the bootloader mode is
+exclusively used to update the device firmware. It’s common for these
+devices to require some operation from fwupd to switch the booting mode
+and then to need a reset to enter bootloader mode. Often, the device is
+enumerated differently in both modes, so fwupd needs to know that the
+same device will be identified differently depending on the boot mode.
+
+The common way to do this is to add the
+`FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG` flag in the device before its detach
+method returns. This will make fwupd wait for a predetermined amount of
+time for the device to be detected again. Then, to inform fwupd about
+the two identities of the same device, the CounterpartGuid key can be
+used in a device entry to match it with another defined device (example:
+https://github.com/fwupd/fwupd/blob/main/plugins/steelseries/steelseries.quirk).
+
+### Inhibiting a device
+
+If a device becomes unsuitable for an update for whatever reason (see
+“Battery checks” above for an example), a plugin can temporarily disable
+firmware updates on it by calling [fu_device_inhibit()](https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-device.c). The device will
+still be listed as present by `fwupdmgr get-devices`, but fwupd won’t
+allow firmware updates on it.  Device inhibition can be disabled with
+[fu_device_uninhibit()](https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-device.c).
```

Comment:
maybe add that there can be multiple inhibits on a specific device, and that it's only updatable when all of them are removed.